File: Samples/Docking/VB/DockSample/XamlWriter.NamespaceCache.vb
Imports System
Imports System.Diagnostics
Imports System.Collections.Generic
Imports System.Windows.Markup
Imports System.Globalization
Imports System.Reflection

Partial Class XamlWriter

    Private Class NamespaceCache

        Public Const XamlNamespace As String = "http://schemas.microsoft.com/winfx/2006/xaml"

        Private _defaultNamespace As String
        Private _prefixes As Dictionary(Of String, String) = New Dictionary(Of String, String)  ' Key: namespace, Value: prefix
        Private _assemblyNamespaces As Dictionary(Of String, Dictionary(Of String, String)) = New Dictionary(Of String, Dictionary(Of String, String))

        Private ReadOnly Property DefaultNamespace() As String
            Get
                Return _defaultNamespace
            End Get
        End Property

        Public Function GetPrefixForNamespace(ByVal ns As String) As String
            Dim ret As String = String.Empty
            If ns <> DefaultNamespace AndAlso _prefixes.ContainsKey(ns) Then
                ret = _prefixes(ns)
            End If
            Return ret
        End Function

        Public Function GetNamespace(ByVal type As Type) As String
            Dim uri As String = EnsureNamespaceMaps(type)
            If (_defaultNamespace Is Nothing) Then
                _defaultNamespace = uri
            End If
            If Not _prefixes.ContainsKey(uri) Then
                Dim prefix As String = ("ns" + (_prefixes.Count + 1).ToString(CultureInfo.InvariantCulture))
                _prefixes(uri) = prefix
            End If
            Return uri
        End Function

        Private Function EnsureNamespaceMaps(ByVal type As Type) As String
            Dim assemblyFullName As String = type.Assembly.FullName
            Dim namespaces As Dictionary(Of String, String)
            If _assemblyNamespaces.ContainsKey(assemblyFullName) Then
                namespaces = _assemblyNamespaces(assemblyFullName)
            Else
                For Each prefixAttribute As XmlnsPrefixAttribute In type.Assembly.GetCustomAttributes(GetType(XmlnsPrefixAttribute), True)
                    _prefixes(prefixAttribute.XmlNamespace) = prefixAttribute.Prefix
                Next
                namespaces = New Dictionary(Of String, String)
                _assemblyNamespaces.Add(assemblyFullName, namespaces)
                For Each definitionAttribute As XmlnsDefinitionAttribute In type.Assembly.GetCustomAttributes(GetType(XmlnsDefinitionAttribute), True)
                    If (definitionAttribute.AssemblyName Is Nothing) Then
                        namespaces(definitionAttribute.ClrNamespace) = definitionAttribute.XmlNamespace
                    End If
                Next
            End If
            If Not namespaces.ContainsKey(type.Namespace) Then
                Dim uri As String = String.Format(CultureInfo.InvariantCulture, "clr-namespace:{0};assembly={1}", type.Namespace, GetAssemblyNameFromType(type))
                namespaces.Add(type.Namespace, uri)
            End If
            Return namespaces(type.Namespace)
        End Function

        Public Shared Function GetAssemblyNameFromType(ByVal type As Type) As String
            Dim names() As String = type.Assembly.FullName.Split(",")
            If names.Length > 0 Then
                Return names(0)
            Else
                Return String.Empty
            End If
        End Function
    End Class
End Class

File: Samples/Docking/CSharp/DockSample/XamlWriter.NamespaceCache.cs
using System;
using System.Diagnostics;
using System.Collections.Generic;
using System.Windows.Markup;
using System.Globalization;
using System.Reflection;

namespace DevZest.DockSample
{
    partial class XamlWriter
    {
        private class NamespaceCache
        {
            public const string XamlNamespace = "http://schemas.microsoft.com/winfx/2006/xaml";

            private string _defaultNamespace;
            private Dictionary<string, string> _prefixes = new Dictionary<string, string>();  // Key: namespace, Value: prefix
            private Dictionary<string, Dictionary<string, string>> _assemblyNamespaces = new Dictionary<string, Dictionary<string, string>>();

            private string DefaultNamespace
            {
                get { return _defaultNamespace; }
            }

            public string GetPrefixForNamespace(string ns)
            {
                string ret = string.Empty;
                if (ns != DefaultNamespace && _prefixes.ContainsKey(ns))
                    ret = _prefixes[ns];
                return ret;
            }

            public string GetXmlNamespace(Type type)
            {
                string uri = EnsureNamespaceMaps(type);
                if (_defaultNamespace == null)
                    _defaultNamespace = uri;

                if (!_prefixes.ContainsKey(uri))
                {
                    string prefix = "ns" + (_prefixes.Count + 1).ToString(CultureInfo.InvariantCulture);
                    _prefixes[uri] = prefix;
                }
                return uri;
            }

            private string EnsureNamespaceMaps(Type type)
            {
                string assemblyFullName = type.Assembly.FullName;
                Dictionary<string, string> namespaces;
                if (_assemblyNamespaces.ContainsKey(assemblyFullName))
                    namespaces = _assemblyNamespaces[assemblyFullName];
                else
                {
                    foreach (XmlnsPrefixAttribute prefixAttribute in type.Assembly.GetCustomAttributes(typeof(XmlnsPrefixAttribute), true))
                        _prefixes[prefixAttribute.XmlNamespace] = prefixAttribute.Prefix;

                    namespaces = new Dictionary<string, string>();
                    _assemblyNamespaces.Add(assemblyFullName, namespaces);
                    foreach (XmlnsDefinitionAttribute definitionAttribute in type.Assembly.GetCustomAttributes(typeof(XmlnsDefinitionAttribute), true))
                    {
                        if (definitionAttribute.AssemblyName == null)
                            namespaces[definitionAttribute.ClrNamespace] = definitionAttribute.XmlNamespace;
                    }
                }

                if (!namespaces.ContainsKey(type.Namespace))
                {
                    string uri = string.Format(CultureInfo.InvariantCulture, "clr-namespace:{0};assembly={1}", type.Namespace, GetAssemblyNameFromType(type));
                    namespaces.Add(type.Namespace, uri);
                }

                return namespaces[type.Namespace];
            }

            public static string GetAssemblyNameFromType(Type type)
            {
                string[] names = type.Assembly.FullName.Split(',');
                return names.Length > 0 ? names[0] : string.Empty;
            }
        }
    }
}
